### PR TITLE
Function getOptimalPosition should prefer positions inside viewport a…

### PR DESCRIPTION
**Pull request by niegowski**
…nd not lock on first not optimal position

(because of intersection flipping between viewport and limiter for different position).

### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Balloon panels should prefer positions inside viewport . Closes ckeditor/ckeditor5#6181.

---

### CODE REVIEW

**Comment by niegowski** [comment]
cc @ckeditor/qa-team  Could you test this PR? This function is used for positioning:
* Button of BlockToolbar, 
* Panel of DropdownView, 
* BalloonPanelView.

**Comment by FilipTokarski** [comment]
Tested in all browsers, looks good 👍

File: tests/dom/position.js [hunk 1]
```diff
@@ -272,11 +327,30 @@ describe( 'getOptimalPosition()', () => {
...
+				positions: [ attachRightBottom, attachLeftBottom ]
 			}, {
 				top: 100,
 				left: 10,
-				name: 'right'
+				name: 'right-bottom'
+			} );
+		} );
+
+		it( 'should return first position that completely fits limiter', () => {
```
Comment (oleq) [inline tests/dom/position.js:338]:
```suggestion
		it( 'should return the first position that completely fits in the limiter', () => {
```

File: tests/dom/position.js [hunk 2]
```diff
@@ -402,13 +476,140 @@ describe( 'getOptimalPosition()', () => {
...
 				fitInViewport: true
 			}, {
 				top: 100,
 				left: 10,
-				name: 'right'
+				name: 'right-bottom'
+			} );
+		} );
+
+		it( 'should prefer position fully fitting in viewport', () => {
```
Comment (oleq) [inline tests/dom/position.js:488]:
```suggestion
		it( 'should prefer positions fitting entirely into the viewport', () => {
```

File: tests/dom/position.js [hunk 3]
```diff
@@ -402,13 +476,140 @@ describe( 'getOptimalPosition()', () => {
...
+				left: 35,
+				name: 'right-bottom'
+			} );
+		} );
+	} );
+
+	describe( 'maximizes intersection area with both limiter and viewport', () => {
+		beforeEach( setElementTargetBigLimiterPlayground );
+
+		it( 'should prefer position with bigger intersection (#1)', () => {
```
Comment (oleq) [inline tests/dom/position.js:512]:
```suggestion
		it( 'should prefer a position with a bigger intersection area (#1)', () => {
```

File: tests/dom/position.js [hunk 4]
```diff
@@ -402,13 +476,140 @@ describe( 'getOptimalPosition()', () => {
...
+				height: 20
+			} );
+			assertPositionName( {
+				element, target, limiter,
+				positions: allPositions,
+				fitInViewport: true
+			}, 'right-bottom' );
+		} );
+
+		it( 'should prefer position with bigger intersection (#2)', () => {
```
Comment (oleq) [inline tests/dom/position.js:528]:
```suggestion
		it( 'should prefer a position with a bigger intersection area (#2)', () => {
```

File: tests/dom/position.js [hunk 5]
```diff
@@ -402,13 +476,140 @@ describe( 'getOptimalPosition()', () => {
...
+				height: 20
+			} );
+			assertPositionName( {
+				element, target, limiter,
+				positions: allPositions,
+				fitInViewport: true
+			}, 'right-top' );
+		} );
+
+		it( 'should prefer position with bigger intersection (#3)', () => {
```
Comment (oleq) [inline tests/dom/position.js:544]:
```suggestion
		it( 'should prefer a position with a bigger intersection area (#3)', () => {
```

File: tests/dom/position.js [hunk 6]
```diff
@@ -402,13 +476,140 @@ describe( 'getOptimalPosition()', () => {
...
+				height: 20
+			} );
+			assertPositionName( {
+				element, target, limiter,
+				positions: allPositions,
+				fitInViewport: true
+			}, 'left-bottom' );
+		} );
+
+		it( 'should prefer position with bigger intersection (#4)', () => {
```
Comment (oleq) [inline tests/dom/position.js:560]:
```suggestion
		it( 'should prefer a position with a bigger intersection area (#4)', () => {
```

File: tests/dom/position.js [hunk 7]
```diff
@@ -402,13 +476,140 @@ describe( 'getOptimalPosition()', () => {
...
+				height: 20
+			} );
+			assertPositionName( {
+				element, target, limiter,
+				positions: allPositions,
+				fitInViewport: true
+			}, 'left-top' );
+		} );
+
+		it( 'should not stick to first biggest intersection in one area', () => {
```
Comment (oleq) [inline tests/dom/position.js:576]:
```suggestion
		it( 'should not stick to the first biggest intersection in one area', () => {
```

File: tests/dom/position.js [hunk 8]
```diff
@@ -402,13 +476,140 @@ describe( 'getOptimalPosition()', () => {
...
+				fitInViewport: true
+			}, {
+				top: 100,
+				left: 35,
+				name: 'right-bottom'
+			} );
+		} );
+	} );
+
+	describe( 'maximizes intersection area with both limiter and viewport', () => {
```
Comment (oleq) [inline tests/dom/position.js:509]:
```suggestion
	describe( 'optimisation in the context of both the limiter and the viewport', () => {
```

**Pull request merged**